lib/senders/index.js

```diff
@@ -24,143 +21,228 @@ module.exports = function (config, log) {
-            uaOS: opts.uaOS,
-            uaOSVersion: opts.uaOSVersion
-          }
-        ))
-      }
-      mailer.sendPostVerifyEmail = function (email, opts) {
-        return P.resolve(mailer.postVerifyEmail(
-          {
-            email: email,
-            acceptLanguage: opts.acceptLanguage || defaultLanguage
+  .then(function (senders) {
+    var ungatedMailer = senders.email
+    var configBounces = config.bounces || {}
+    var MAX_HARD = configBounces.hard && configBounces.hard.max || 0
+    var MAX_SOFT = configBounces.soft && configBounces.soft.max || 0
+    var MAX_COMPLAINT = configBounces.complaint && configBounces.complaint.max || 0
+    var DURATION_HARD = configBounces.hard && configBounces.hard.duration || Infinity
+    var DURATION_SOFT = configBounces.soft && configBounces.soft.duration || Infinity
+    var DURATION_COMPLAINT = configBounces.complaint && configBounces.complaint.duration || Infinity
+    var BOUNCE_TYPE_HARD = 1
+    var BOUNCE_TYPE_SOFT = 2
+    var BOUNCE_TYPE_COMPLAINT = 3
+
+    function checkBounce(bounce, rules, now) {
+      for (var i = 0; i < rules.length; i++) {
+        var ruleSet = rules[i]
+
+        if (bounce.bounceType === ruleSet.type) {
+          if (bounce.createdAt > now - ruleSet.duration) {
+            ruleSet.count++
+            if (ruleSet.count > ruleSet.max) {
+              throw ruleSet.error()
+            }
           }
-        ))
+          break
+        }
       }
-      mailer.sendUnblockCode = function (account, unblockCode, opts) {
-        return P.resolve(mailer.unblockCodeEmail(
-          {
-            acceptLanguage: opts.acceptLanguage || defaultLanguage,
-            flowId: opts.flowId,
-            flowBeginTime: opts.flowBeginTime,
-            email: account.email,
-            ip: opts.ip,
-            location: opts.location,
-            timeZone: opts.timeZone,
-            uaBrowser: opts.uaBrowser,
-            uaBrowserVersion: opts.uaBrowserVersion,
-            uaOS: opts.uaOS,
-            uaOSVersion: opts.uaOSVersion,
-            uid: account.uid.toString('hex'),
-            unblockCode: unblockCode
-          }
-        ))
+    }
+
+    var bounceRules = [
+      {
+        count: 0,
+        duration: DURATION_HARD,
+        error: error.emailBouncedHard,
+        max: MAX_HARD,
+        type: BOUNCE_TYPE_HARD
+      },
+      {
+        count: 0,
+        duration: DURATION_COMPLAINT,
+        error: error.emailComplaint,
+        max: MAX_COMPLAINT,
+        type: BOUNCE_TYPE_COMPLAINT
+      },
+      {
+        count: 0,
+        duration: DURATION_SOFT,
+        error: error.emailBouncedSoft,
+        max: MAX_SOFT,
+        type: BOUNCE_TYPE_SOFT
       }
-      return senders
+    ]
+
+    function bounceGatedMailer(email) {
+      return db.emailBounces(email)
+        .then(function (bounces) {
+          bounceRules[0].count = bounceRules[1].count = bounceRules[2].count= 0
```

Comment:
Using a single shared object for these seems like a potential future footgun, if we e.g. later refactor the checking to be asynchronous and start getting conflicting writes to the object.  Would it make sense to either work on a copy, or store the mutable parts in a separate object that's created afresh each time?